pkg/expressions/function/functions.go

```diff
@@ -769,6 +808,130 @@ func getChartFromDiscoveredArtifacts(artifacts *kargoapi.DiscoveredArtifacts) ex
 	}
 }
 
+// getArtifactFromFreight returns a function that finds an artifact based on the
+// provided subscription name and optional origin.
+//
+// The returned function uses freight requests and references to locate the
+// appropriate artifact within the project context.
+func getArtifactFromFreight(
+	ctx context.Context,
+	c client.Client,
+	project string,
+	freightReqs []kargoapi.FreightRequest,
+	freightRefs []kargoapi.FreightReference,
+) exprFn {
+	return func(a ...any) (any, error) {
+		if len(a) == 0 || len(a) > 2 {
+			return nil, fmt.Errorf("expected 1-2 arguments, got %d", len(a))
+		}
+
+		subName, ok := a[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("first argument must be string, got %T", a[0])
+		}
+
+		var desiredOrigin *kargoapi.FreightOrigin
+		if len(a) == 2 {
+			origin, ok := a[1].(kargoapi.FreightOrigin)
+			if !ok {
+				return nil,
+					fmt.Errorf("second argument must be FreightOrigin, got %T", a[1])
+			}
+			desiredOrigin = &origin
+		}
+
+		artifact, err := freight.FindArtifact(
+			ctx,
+			c,
+			project,
+			freightReqs,
+			desiredOrigin,
+			freightRefs,
+			subName,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error finding artifact for subscription %s: %w", subName, err)
+		}
+		if artifact == nil {
+			return nil, nil
+		}
+
+		// Unpack details into a map[string]any so they're actually accessible from
+		// within an expression.
+		exprArtifact := expressionFriendlyGenericArtifact{
+			SubscriptionName: artifact.SubscriptionName,
+			Version:          artifact.Version,
+			Details:          map[string]any{},
+		}
+		if artifact.Details != nil {
+			if err := json.Unmarshal(
+				artifact.Details.Raw,
+				&exprArtifact.Details,
+			); err != nil {
+				return nil, fmt.Errorf(
+					"error unmarshaling artifact details for subscription %s: %w",
+					subName, err,
+				)
+			}
+		}
+
+		return exprArtifact, nil
+	}
+}
+
+// getArtifactFromDiscoveredArtifacts returns a function that finds an artifact
+// based on the provided subscription name.
+func getArtifactFromDiscoveredArtifacts(
+	artifacts *kargoapi.DiscoveredArtifacts,
+) exprFn {
+	return func(a ...any) (any, error) {
+		if len(a) != 1 {
+			return nil, fmt.Errorf("expected 1 argument, got %d", len(a))
+		}
+
+		subName, ok := a[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("first argument must be string, got %T", a[0])
+		}
+
+		if artifacts == nil {
+			return nil, nil
+		}
+
+		var artifact *kargoapi.GenericArtifactReference
+		for _, result := range artifacts.OtherResults {
+			if result.SubscriptionName != subName {
+				continue
+			}
+			if len(result.ArtifactReferences) > 0 {
+				artifact = &result.ArtifactReferences[0]
+			}
+		}
+
+		// Unpack details into a map[string]any so they're actually accessible from
```

Comment:
Nit: This requirement should probably be documented in the `Details` field for clarity

Reply:
Also realized that now that details has changed to "metadata," there might be another small tweak needed to these functions.

Will fix both.

Reply:
Fixed in b79fcc5cd58ff1eb437a0e1ad8ca440619218318